test/versioned/fastify/add-hook.tap.js

```diff
@@ -0,0 +1,144 @@
+
+// these events fire after the route
+// handler. they are in separate arrays
+// for segment relationship assertions later
+const AFTER_HANDLER_HOOKS = ['preSerialization', 'onSend']
+
+// the onResponse hook fires after a response
+// is received by client which is out of context
+// of the transaction
+const AFTER_TX_HOOKS = ['onResponse']
+
+const ALL_HOOKS = [...REQUEST_HOOKS, ...AFTER_HANDLER_HOOKS, ...AFTER_TX_HOOKS]
+
+function getSegmentNames(hooks) {
+  return hooks.map((hookName) => {
+    return `Nodejs/Middleware/Fastify/${hookName}/testHook`
+  })
+}
+
+tap.test('fastify hook instrumentation', (t) => {
+  t.autoend()
+
+  let agent = null
+  let fastify = null
+
+  t.beforeEach(() => {
+    agent = helper.instrumentMockedAgent({
+      feature_flag: {
+        fastify_instrumentation: true
+      }
+    })
+    fastify = require('fastify')()
+  })
+
+  t.afterEach(() => {
+    helper.unloadAgent(agent)
+    fastify.close()
+  })
+
+  t.test('non-error hooks', async function nonErrorHookTest(t) {
+    // setup fastify route
+    fastify.get('/', function routeHandler(_, reply) {
+      t.comment('route handler called')
+      reply.send({ hello: 'world' })
+    })
+
+    // setup hooks
+    const ok = ALL_HOOKS.reduce((all, hookName) => {
+      t.comment(`registering ${hookName}`)
+      all[hookName] = false
+      return all
+    }, {})
+
+    ALL_HOOKS.forEach((hookName) => {
+      fastify.addHook(hookName, function testHook(...args) {
+        t.comment(`${hookName} called`)
+        // lifecycle signatures vary between the events
+        // the last arg is always the next function though
+        const next = args[args.length - 1]
+        ok[hookName] = true
+        next()
+      })
+    })
+
+    agent.on('transactionFinished', (transaction) => {
+      t.equal('WebFrameworkUri/Fastify/GET//', transaction.getName(), `transaction name matched`)
+      // all the hooks are siblings of the route handler
+      // except the AFTER_HANDLER_HOOKS which are children of the route handler
+      metrics.assertSegments(transaction.trace.root, [
+        'WebTransaction/WebFrameworkUri/Fastify/GET//',
+        [
+          ...getSegmentNames(REQUEST_HOOKS),
+          'Nodejs/Middleware/Fastify/routeHandler',
+          getSegmentNames(AFTER_HANDLER_HOOKS)
+        ]
+      ])
+    })
+
+    await fastify.listen(0)
+    const { port } = fastify.server.address()
+    const result = await getAsync(`http://127.0.0.1:${port}/`)
+    t.equal(result.statusCode, 200)
+    t.equal(result.body, JSON.stringify({ hello: 'world' }))
+
+    // verify every hook was called after response
+    for (const [hookName, isOk] of Object.entries(ok)) {
+      t.equal(isOk, true, `${hookName} captured`)
+    }
+    t.end()
+  })
+
+  t.test('error hook', async function errorHookTest(t) {
+    // setup fastify route
+    fastify.get('/', function routeHandler() {
+      throw new Error('sup')
```

Comment:
I think because in fastify 2 it didn't handle throwing errors in synchronous handlers.  So just turns this handler into an async function and it will work. It's mentioned in the 2 -> 3 upgrade guide https://github.com/fastify/fastify/blob/main/docs/Migration-Guide-V3.md#manage-uncaught-exception-2073